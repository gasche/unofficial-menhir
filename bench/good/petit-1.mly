(* This file defines the parser. The set of tokens that it exploits
   is defined in file [tokens.mly]. *)

%{

  open MIPSOps
  open LPP
  open Primitive

%}

%start <LPP.program> program

(* Tokens (or pseudo-tokens, used in %prec declarations) with lower
   precedence come first in this list. *)

%left OR
%left AND
%nonassoc NOT
%nonassoc LT LE GT GE EQ NE
%left MINUS PLUS
%left TIMES SLASH
%nonassoc unary_minus
%nonassoc LBRACKET
%nonassoc THEN
%nonassoc ELSE

%%

(* Formal and actual parameter lists are delimited with parentheses
   and separated with semicolons and commas, respectively. Contrary to
   standard Pascal, the parentheses are not optional (it would then
   become impossible to distinguish a variable from a function call
   with no parameters). *)

(* ------------------------------------------------------------------------- *)
(* Types. *)

typ:
| INTEGER
    { TypInt }
| BOOLEAN
    { TypBool }
| ARRAY OF t = typ
    { TypArray t }

(* ------------------------------------------------------------------------- *)
(* Expressions. *)

expression:
  e = raw_expression
    { Location.make $startpos $endpos e }
| LPAREN e = expression RPAREN
    { e }

raw_expression:
  i = INTCONST
    { EConst (ConstInt i) }
| b = BOOLCONST
    { EConst (ConstBool b) }
| id = ID
    { EGetVar id }
| MINUS e = expression %prec unary_minus
    { EUnOp (UOpNeg, e) }
| e1 = expression op = binop e2 = expression
    { EBinOp (op, e1, e2) }
| c = callee LPAREN actuals = separated_list(COMMA, expression) RPAREN
    { EFunCall (Location.make $startpos(c) $endpos(c) c, actuals) }
| a = expression LBRACKET i = expression RBRACKET
    { EArrayGet (a, i) }
| NEW ARRAY OF t = typ LBRACKET e = expression RBRACKET
    { EArrayAlloc (t, e) }

(* Binary operators are defined as a separate nonterminal symbol
   for enhanced readability. However, note that this symbol is
   declared %inline, which means that its definition is in fact
   textually expanded into the definition of expressions above.
   This is necessary for operator priorities to be properly
   taken into account. *)

%inline binop:
| PLUS
    { OpAdd }
| MINUS
    { OpSub }
| TIMES
    { OpMul }
| SLASH
    { OpDiv }
| LT
    { OpLt }
| LE
    { OpLe }
| GT
    { OpGt }
| GE
    { OpGe }
| EQ
    { OpEq }
| NE
    { OpNe }

(* ------------------------------------------------------------------------- *)
(* Conditions.

   We first define ``nontrivial'' conditions, which cannot simply be an
   expression or a parenthesized expression: they must involve at least
   one of the Boolean operators NOT, AND, OR. Then, we define a condition
   to be either a (Boolean-valued) expression or a nontrivial condition.

   This two-step approach is necessary to avoid ambiguity with
   parentheses. Indeed, conditions include expressions, and both
   conditions and expressions can be parenthesized, so a naïve
   approach would yield an ambiguous grammar. Thanks to Yannick Moy
   for suggesting this solution. *)

nontrivial_condition:
| NOT c = condition
    { CNot c }
| c1 = condition AND c2 = condition
    { CAnd (c1, c2) }
| c1 = condition OR c2 = condition
    { COr (c1, c2) }
| LPAREN c = nontrivial_condition RPAREN
    { c }

condition:
| e = expression
    { CExpression e }
| c = nontrivial_condition
    { c }

(* ------------------------------------------------------------------------- *)
(* Instructions and blocks. *)

lvalue:
| id = ID
    { fun e -> ISetVar (id, e) }
| a = expression LBRACKET i = expression RBRACKET
    { fun e -> IArraySet (a, i, e) }

(* The explicit type constraint on [lvalue] below is redundant, but
   serves to avoid an ocaml warning about the code generated by
   Menhir. The need for this constraint should disappear with future
   versions of Menhir. *)

instruction:
| c = callee LPAREN actuals = separated_list(COMMA, expression) RPAREN
    { IProcCall (Location.make $startpos(c) $endpos(c) c, actuals) }
| lvalue = lvalue COLONEQ e = expression
    { (lvalue : LPP.expression -> LPP.instruction) e }
| READLN LPAREN lvalue = lvalue RPAREN
    { let callee = Location.make $startpos($1) $endpos($1) (CPrimitiveFunction Readln) in
      let e = Location.make $startpos $endpos (EFunCall (callee, [])) in
      (lvalue : LPP.expression -> LPP.instruction) e }
| IF c = condition THEN b = instruction_or_block
    { IIf (c, b, ISeq []) }
| IF c = condition THEN b1 = instruction_or_block ELSE b2 = instruction_or_block
    { IIf (c, b1, b2) }
| WHILE c = condition DO b = instruction_or_block
    { IWhile (c, b) }

block:
| BEGIN is = separated_list(SEMICOLON, instruction) END
    { ISeq is }

instruction_or_block:
| i = instruction
    { i }
| b = block
    { b }

(* ------------------------------------------------------------------------- *)
(* Callees. *)

callee:
| WRITE
    { CPrimitiveFunction Write }
| WRITELN
    { CPrimitiveFunction Writeln }
| id = ID
    { CUserFunction (Location.content id)}

(* ------------------------------------------------------------------------- *)
(* Procedures and functions. *)

(* Procedures and functions are very similar, so we are able to
   express a single semantic action that covers both productions. In
   the case of procedures, the variable [result] is bound to the value
   [None] via a simple trick: we recognize the non-terminal
   [no_result_type], whose language is empty. *)

procedure:
| FUNCTION f = ID
  LPAREN formals = separated_bindings RPAREN
  COLON result = some_result_type SEMICOLON
  locals = variables
  body = block
  SEMICOLON

| PROCEDURE f = ID
  LPAREN formals = separated_bindings RPAREN
  result = no_result_type SEMICOLON
  locals = variables
  body = block
  SEMICOLON

    {
      f, {
        formals = formals;
        result = result;
        locals = locals;
        body = body
      }
    }

%inline no_result_type:
  (* nothing *)
    { None }

%inline some_result_type:
  t = typ
    { Some t }

separated_bindings:
| bindings = separated_list(SEMICOLON, binding) (* list can be empty *)
    { List.flatten bindings }

terminated_bindings:
| bindings = terminated(binding, SEMICOLON)+    (* list is nonempty *)
    { List.flatten bindings }

binding:
| ids = separated_nonempty_list(COMMA, ID) COLON t = typ
    { List.map (fun id -> (id, t)) ids }

variables:
| vars = loption(preceded(VAR, terminated_bindings))
    { vars }

(* ------------------------------------------------------------------------- *)
(* Programs. *)

program:
  PROGRAM
  globals = variables
  defs = procedure*
  main = block
  DOT
    {{
       globals = globals;
       defs = defs;
       main = main
    }}

