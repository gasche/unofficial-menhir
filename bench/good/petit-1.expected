Warning: you are using the standard library and/or the %inline keyword. We
recommend switching on --infer in order to avoid obscure type error messages.
%{

  open MIPSOps
  open LPP
  open Primitive

%}
%start program
%token WRITELN
%token WRITE
%token WHILE
%token VAR
%token SEMICOLON
%token RPAREN
%token READLN
%token RBRACKET
%token PROGRAM
%token PROCEDURE
%token OF
%token NEW
%token LPAREN
%token INTEGER
%token <int32> INTCONST
%token IF
%token <string Location.t> ID
%token FUNCTION
%token END
%token DOT
%token DO
%token COMMA
%token COLONEQ
%token COLON
%token BOOLEAN
%token <bool> BOOLCONST
%token BEGIN
%token ARRAY
%token OR
%token AND
%token NOT
%token NE
%token LT
%token LE
%token GT
%token GE
%token EQ
%token PLUS
%token MINUS
%token TIMES
%token SLASH
%token LBRACKET
%token THEN
%token ELSE
%left OR 
%left AND 
%nonassoc NOT 
%nonassoc NE LT LE GT GE EQ 
%left PLUS MINUS 
%left TIMES SLASH 
%nonassoc unary_minus 
%nonassoc LBRACKET 
%nonassoc THEN 
%nonassoc ELSE 
%type <LPP.program> program
%%

typ:
| INTEGER
    { TypInt }
| BOOLEAN
    { TypBool }
| ARRAY OF t = typ
    { TypArray t }

expression:
| e = raw_expression
    { Location.make $startpos $endpos e }
| LPAREN e = expression RPAREN
    { e }

raw_expression:
| i = INTCONST
    { EConst (ConstInt i) }
| b = BOOLCONST
    { EConst (ConstBool b) }
| id = ID
    { EGetVar id }
| MINUS e = expression %prec unary_minus
    { EUnOp (UOpNeg, e) }
| e1 = expression PLUS e2 = expression
    {let op =
  ( OpAdd )
in
 EBinOp (op, e1, e2) }
| e1 = expression MINUS e2 = expression
    {let op =
  ( OpSub )
in
 EBinOp (op, e1, e2) }
| e1 = expression TIMES e2 = expression
    {let op =
  ( OpMul )
in
 EBinOp (op, e1, e2) }
| e1 = expression SLASH e2 = expression
    {let op =
  ( OpDiv )
in
 EBinOp (op, e1, e2) }
| e1 = expression LT e2 = expression
    {let op =
  ( OpLt )
in
 EBinOp (op, e1, e2) }
| e1 = expression LE e2 = expression
    {let op =
  ( OpLe )
in
 EBinOp (op, e1, e2) }
| e1 = expression GT e2 = expression
    {let op =
  ( OpGt )
in
 EBinOp (op, e1, e2) }
| e1 = expression GE e2 = expression
    {let op =
  ( OpGe )
in
 EBinOp (op, e1, e2) }
| e1 = expression EQ e2 = expression
    {let op =
  ( OpEq )
in
 EBinOp (op, e1, e2) }
| e1 = expression NE e2 = expression
    {let op =
  ( OpNe )
in
 EBinOp (op, e1, e2) }
| c = callee LPAREN xs0 = loption_separated_nonempty_list_COMMA_expression__ RPAREN
    {let actuals =
  let xs = xs0 in
  ( xs )
in
 EFunCall (Location.make $startpos(c) $endpos(c) c, actuals) }
| a = expression LBRACKET i = expression RBRACKET
    { EArrayGet (a, i) }
| NEW ARRAY OF t = typ LBRACKET e = expression RBRACKET
    { EArrayAlloc (t, e) }

condition:
| e = expression
    { CExpression e }
| c = nontrivial_condition
    { c }

lvalue:
| id = ID
    { fun e -> ISetVar (id, e) }
| a = expression LBRACKET i = expression RBRACKET
    { fun e -> IArraySet (a, i, e) }

procedure:
| FUNCTION f = ID LPAREN formals = separated_bindings RPAREN COLON t0 = typ SEMICOLON locals = variables body = block SEMICOLON
    {let result =
  let t = t0 in
  ( Some t )
in

      f, {
        formals = formals;
        result = result;
        locals = locals;
        body = body
      }
    }
| PROCEDURE f = ID LPAREN formals = separated_bindings RPAREN SEMICOLON locals = variables body = block SEMICOLON
    {let result =
  ( None )
in

      f, {
        formals = formals;
        result = result;
        locals = locals;
        body = body
      }
    }

separated_bindings:
| xs0 = loption_separated_nonempty_list_SEMICOLON_binding__
    {let bindings =
  let xs = xs0 in
  ( xs )
in
 List.flatten bindings }

terminated_bindings:
| bindings = nonempty_list_terminated_binding_SEMICOLON__
    { List.flatten bindings }

variables:
| vars = loption_preceded_VAR_terminated_bindings__
    { vars }

program:
| PROGRAM globals = variables defs = list_procedure_ main = block DOT
    {{
       globals = globals;
       defs = defs;
       main = main
    }}

loption_separated_nonempty_list_SEMICOLON_instruction__:
| 
    { [] }
| x = separated_nonempty_list_SEMICOLON_instruction_
    { x }

loption_separated_nonempty_list_SEMICOLON_binding__:
| 
    { [] }
| x = separated_nonempty_list_SEMICOLON_binding_
    { x }

loption_separated_nonempty_list_COMMA_expression__:
| 
    { [] }
| x = separated_nonempty_list_COMMA_expression_
    { x }

loption_preceded_VAR_terminated_bindings__:
| 
    { [] }
| VAR x0 = terminated_bindings
    {let x =
  let x = x0 in
  ( x )
in
 x }

list_procedure_:
| 
    { [] }
| x = procedure xs = list_procedure_
    { x :: xs }

nonempty_list_terminated_binding_SEMICOLON__:
| x0 = binding SEMICOLON
    {let x =
  let x = x0 in
  ( x )
in
 [ x ] }
| x0 = binding SEMICOLON xs = nonempty_list_terminated_binding_SEMICOLON__
    {let x =
  let x = x0 in
  ( x )
in
 x :: xs }

separated_nonempty_list_SEMICOLON_instruction_:
| x = instruction
    { [ x ] }
| x = instruction SEMICOLON xs = separated_nonempty_list_SEMICOLON_instruction_
    { x :: xs }

separated_nonempty_list_SEMICOLON_binding_:
| x = binding
    { [ x ] }
| x = binding SEMICOLON xs = separated_nonempty_list_SEMICOLON_binding_
    { x :: xs }

separated_nonempty_list_COMMA_expression_:
| x = expression
    { [ x ] }
| x = expression COMMA xs = separated_nonempty_list_COMMA_expression_
    { x :: xs }

separated_nonempty_list_COMMA_ID_:
| x = ID
    { [ x ] }
| x = ID COMMA xs = separated_nonempty_list_COMMA_ID_
    { x :: xs }

nontrivial_condition:
| NOT c = condition
    { CNot c }
| c1 = condition AND c2 = condition
    { CAnd (c1, c2) }
| c1 = condition OR c2 = condition
    { COr (c1, c2) }
| LPAREN c = nontrivial_condition RPAREN
    { c }

instruction:
| c = callee LPAREN xs0 = loption_separated_nonempty_list_COMMA_expression__ RPAREN
    {let actuals =
  let xs = xs0 in
  ( xs )
in
 IProcCall (Location.make $startpos(c) $endpos(c) c, actuals) }
| lvalue = lvalue COLONEQ e = expression
    { (lvalue : LPP.expression -> LPP.instruction) e }
| READLN LPAREN lvalue = lvalue RPAREN
    { let callee = Location.make $startpos($1) $endpos($1) (CPrimitiveFunction Readln) in
      let e = Location.make $startpos $endpos (EFunCall (callee, [])) in
      (lvalue : LPP.expression -> LPP.instruction) e }
| IF c = condition THEN b = instruction_or_block
    { IIf (c, b, ISeq []) }
| IF c = condition THEN b1 = instruction_or_block ELSE b2 = instruction_or_block
    { IIf (c, b1, b2) }
| WHILE c = condition DO b = instruction_or_block
    { IWhile (c, b) }

block:
| BEGIN xs0 = loption_separated_nonempty_list_SEMICOLON_instruction__ END
    {let is =
  let xs = xs0 in
  ( xs )
in
 ISeq is }

instruction_or_block:
| i = instruction
    { i }
| b = block
    { b }

callee:
| WRITE
    { CPrimitiveFunction Write }
| WRITELN
    { CPrimitiveFunction Writeln }
| id = ID
    { CUserFunction (Location.content id)}

binding:
| ids = separated_nonempty_list_COMMA_ID_ COLON t = typ
    { List.map (fun id -> (id, t)) ids }

%%



